File: Algorithms/092.ReverseLinkedListII.py
# http://jessicasco.github.io/blog/2015/05/22/python-swap-caution/
class Solution:
    # @param {ListNode} head
    # @param {integer} m
    # @param {integer} n
    # @return {ListNode}
    def reverseBetween(self, head, m, n):
        if m == n:
            return head
        headptr = ListNode(0)
        headptr.next = head
        ins, p = headptr, head
        for i in range(1, n):
            if i < m:
                ins, p = ins.next, p.next
            else:
                #temp = p.next
                #p.next = temp.next
                #temp.next = ins.next
                #ins.next = temp

                #p.next, p.next.next, ins.next =  p.next.next, ins.next, p.next
                p.next.next, p.next,  ins.next =  ins.next, p.next.next, p.next
        return headptr.next

if __name__ == '__main__':
    S = Solution()
    #head = ListNode(3)
    #head.next = ListNode(5)
    head, head.next = ListNode(3), ListNode(5)
    S.reverseBetween(head, 1, 2)

File: Algorithms/092.ReverseLinkedListII.cpp
#include "leetcode.h"
class Solution {
public:
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        if (m == n) 
            return head;
        ListNode *headptr = new ListNode(0);
        headptr->next = head;
        ListNode *ins = headptr;
        ListNode *p = head;
        for (int i=1; i < n; i++) {
            if (i < m) {
                ins = ins->next;
                p = p->next;
            }
            else {
                ListNode *temp = p->next;
                p->next = temp->next;
                temp->next = ins->next;
                ins->next = temp;
            }
        }
        return headptr->next;
    }
};
int main() {
    Solution s;
    ListNode *l1 = new ListNode(1);
    ListNode *l2 = new ListNode(2);
    ListNode *l3 = new ListNode(3);
    ListNode *l4 = new ListNode(4);
    ListNode *l5 = new ListNode(5);
    l1->next = l2;
    l2->next = l3;
    l3->next = l4;
    l4->next = l5;
    l5->next = NULL;
    ListNode *head = s.reverseBetween(l1, 2, 4);
    while (head != NULL) {
        cout << head->val << endl;
        head = head->next;
    }
    return 0;
}
